src/coreclr/vm/interpexec.cpp

```diff
@@ -1081,13 +1120,24 @@ void InterpExecMethod(InterpreterFrame *pInterpreterFrame, InterpMethodContextFr
                     }
                     else
                     {
-                        // At this stage in the implementation, we assume this is pointer to
-                        // interpreter code. In the future, this should probably be tagged pointer
-                        // for interpreter call or normal pointer for JIT/R2R call.
                         targetIp = (const int32_t*)targetMethod;
                     }
                     }
 CALL_TARGET_IP:
+                    // Interpreter-TODO: we need a fast way to check of the targetIp is an interpreter code or not.
+                    // Probably use a tagged pointer for interpreter code and a normal pointer for JIT/R2R code.
+                    EECodeInfo codeInfo((PCODE)targetIp);
+                    if (!codeInfo.IsValid())
+                    {
+                        EEPOLICY_HANDLE_FATAL_ERROR_WITH_MESSAGE(COR_E_EXECUTIONENGINE, W("Attempted to execute native code from interpreter"));
+                    }
+                    else if (codeInfo.GetCodeManager() != ExecutionManager::GetInterpreterCodeManager())
+                    {
+                        MethodDesc *pMD = codeInfo.GetMethodDesc();
```

Comment:
I think it would be great if we didn't have to work with `MethodDesc` at all for the fast invocation path. Also if these transition thunks would be shared per signature and not owned by each method independently, meaning the compiled `targetIp` would be passed explicitly. So a call pseudocode could look like:
```
obtain targetIp;
if (targetIp is not interp)
    callStubInvoke = check call site cache
    if (!callStubInvoke)
         build CallStubGenerator
         set callStubInvoke and write cache
    callStubInvoke->Invoke(targetIp, ...)
else
    interp call 
```